src/content/app/genome-browser/state/browser-location/browserLocationSlice.ts

```diff
@@ -0,0 +1,173 @@
+
+const isServer = typeof window === 'undefined';
+const chrLocations = !isServer ? browserStorageService.getChrLocation() : [];
+
+export type ChrLocation = [string, number, number];
+
+export type ChrLocations = { [genomeId: string]: ChrLocation };
+
+export type BrowserLocationState = Readonly<{
+  chrLocations: ChrLocations; // final location of the browser when user stopped dragging/zooming; used to update the url
+  actualChrLocations: ChrLocations; // transient locations that change while user is dragging or zooming
+  regionEditorActive: boolean;
+  regionFieldActive: boolean;
+  isObjectInDefaultPosition: boolean;
+}>;
+
+export const defaultBrowserLocationState: BrowserLocationState = {
+  chrLocations,
+  actualChrLocations: {},
+  regionEditorActive: false,
+  regionFieldActive: false,
+  isObjectInDefaultPosition: false
+};
+
+export const setActualChrLocation: ActionCreator<
+  ThunkAction<any, any, null, Action<string>>
+> = (chrLocation: ChrLocation) => {
+  return (dispatch: Dispatch, getState: () => RootState) => {
+    const state = getState();
+    const activeGenomeId = getBrowserActiveGenomeId(state);
+    if (!activeGenomeId) {
+      return;
+    }
+    const payload = {
+      [activeGenomeId]: chrLocation
+    };
+
+    dispatch(updateActualChrLocation(payload));
+  };
+};
+
+export const setChrLocation: ActionCreator<
+  ThunkAction<any, any, null, Action<string>>
+> = (chrLocation: ChrLocation) => {
+  return (dispatch: Dispatch, getState: () => RootState) => {
+    const state = getState();
+    const activeGenomeId = getBrowserActiveGenomeId(state);
+    const activeEnsObjectId = getBrowserActiveEnsObjectId(state);
+    const savedChrLocation = getChrLocation(state);
+    if (!activeGenomeId || !activeEnsObjectId) {
+      return;
+    }
+    const payload = {
+      [activeGenomeId]: chrLocation
+    };
+
+    dispatch(updateChrLocation(payload));
+    browserStorageService.updateChrLocation(payload);
+
+    if (!isEqual(chrLocation, savedChrLocation)) {
+      const newUrl = urlFor.browser({
+        genomeId: activeGenomeId,
+        focus: buildFocusIdForUrl(activeEnsObjectId),
+        location: getChrLocationStr(chrLocation)
+      });
+      dispatch(replace(newUrl));
+    }
+  };
+};
+
+const browserLocationSlice = createSlice({
+  name: 'genome-browser-location',
+  initialState: defaultBrowserLocationState as BrowserLocationState,
+  reducers: {
+    setLocationDataFromUrl(state, action: PayloadAction<ParsedUrlPayload>) {
+      const { activeGenomeId, chrLocation } = action.payload;
+      if (chrLocation) {
+        state.chrLocations[activeGenomeId] = chrLocation;
+      } else {
+        state.chrLocations = pickBy(
+          state.chrLocations,
+          (value, key) => key !== activeGenomeId
+        );
+      }
+    },
+    updateChrLocation(state, action: PayloadAction<ChrLocations>) {
+      state.chrLocations = {
+        ...state.chrLocations,
+        ...action.payload
+      };
```

Comment:
`Object.assign(state.chrLocations, action.payload)`?

Reply:
That's done